lucille-core/src/main/java/com/kmwllc/lucille/stage/CopyFields.java

```diff
@@ -1,70 +1,102 @@
+ *   Defaults to false. </li>
  * </ul>
  */
 public class CopyFields extends Stage {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  private static final ObjectMapper mapper = new ObjectMapper();
 
   public static final Spec SPEC = SpecBuilder.stage()
-      .requiredList("source", new TypeReference<List<String>>(){})
-      .requiredList("dest", new TypeReference<List<String>>(){})
-      .optionalString("update_mode").build();
+      .requiredParent("fieldMapping", new TypeReference<Map<String, String>>() {})
+      .optionalString("update_mode")
+      .optionalBoolean("isNested").build();
 
-  private final List<String> sourceFields;
-  private final List<String> destFields;
+  private final Map<String, Object> fieldMapping;
   private final UpdateMode updateMode;
+  private final boolean isNested;
+  private Map<String, String[]> sourceToDestParts = new HashMap<>();
 
   public CopyFields(Config config) {
     super(config);
-    this.sourceFields = config.getStringList("source");
-    this.destFields = config.getStringList("dest");
+    this.fieldMapping = config.getConfig("fieldMapping").root().unwrapped();
     this.updateMode = UpdateMode.fromConfig(config);
+    this.isNested = ConfigUtils.getOrDefault(config, "isNested", false);
   }
 
   @Override
   public void start() throws StageException {
-    StageUtils.validateFieldNumNotZero(sourceFields, "Copy Fields");
-    StageUtils.validateFieldNumNotZero(destFields, "Copy Fields");
-    StageUtils.validateFieldNumsSeveralToOne(sourceFields, destFields, "Copy Fields");
+    if (this.fieldMapping.size() == 0) {
+      throw new StageException("fieldMapping must have at least one source-dest pair for Copy Fields");
+    }
+
+    if (this.isNested && this.updateMode != null) {
+      log.info("Cannot set both isNested and update_mode in Copy Fields at the same time. Ignoring update_mode, fields will be overwritten.");
+    }
+
+    // create set of destination field parts so we don't have to split them up for every doc
+    for (Entry<String, Object> fieldPair : this.fieldMapping.entrySet()) {
+      String source = fieldPair.getKey();
+      String[] dest = ((String) fieldPair.getValue()).split("\\.");
+      this.sourceToDestParts.put(source, dest);
+    }
   }
 
   @Override
   public Iterator<Document> processDocument(Document doc) throws StageException {
-    for (int i = 0; i < sourceFields.size(); i++) {
-      // If there is only one source or dest, use it. Otherwise, use the current source/dest.
-      String sourceField = sourceFields.get(i);
-      String destField = destFields.size() == 1 ? destFields.get(0) : destFields.get(i);
+    for (Entry<String, String[]> fieldPair : this.sourceToDestParts.entrySet()) {
+      String source = fieldPair.getKey();
+      String[] dest = fieldPair.getValue();
 
-      if (!doc.has(sourceField)) {
-        continue;
+      if (!this.isNested) {
```

Comment:
you don't have to use `this.` before instance variable all the time
it's a convention to do that in the constructor because constructors often take arguments that have the same name as the instance variable
inside a regular method body, you can use `this.` to disambiguate, if the method happens to take an arg with the same name as an instance variable, but it's generally easier to name the method argument so that this doesn't happen